lib/Adapter/Tolerant/Indexer/MemberIndexer.php

```diff
@@ -0,0 +1,149 @@
+<?php
+
+namespace Phpactor\Indexer\Adapter\Tolerant\Indexer;
+
+use Microsoft\PhpParser\Node;
+use Microsoft\PhpParser\Node\Expression\CallExpression;
+use Microsoft\PhpParser\Node\Expression\MemberAccessExpression;
+use Microsoft\PhpParser\Node\Expression\ScopedPropertyAccessExpression;
+use Microsoft\PhpParser\Node\Expression\Variable;
+use Microsoft\PhpParser\Node\QualifiedName;
+use Microsoft\PhpParser\Node\Statement\ExpressionStatement;
+use Microsoft\PhpParser\Token;
+use Phpactor\Indexer\Adapter\Tolerant\TolerantIndexer;
+use Phpactor\Indexer\Model\Index;
+use Phpactor\Indexer\Model\MemberReference;
+use Phpactor\Indexer\Model\RecordReference;
+use Phpactor\Indexer\Model\Record\FileRecord;
+use Phpactor\Indexer\Model\Record\MemberRecord;
+use SplFileInfo;
+
+class MemberIndexer implements TolerantIndexer
+{
+    public function canIndex(Node $node): bool
+    {
+        return $node instanceof ScopedPropertyAccessExpression || $node instanceof MemberAccessExpression;
+    }
+
+    public function index(Index $index, SplFileInfo $info, Node $node): void
+    {
+        if ($node instanceof ScopedPropertyAccessExpression) {
+            $this->indexScopedPropertyAccess($index, $info, $node);
+            return;
+        }
+
+        if ($node instanceof MemberAccessExpression) {
+            $this->indexMemberAccessExpression($index, $info, $node);
+            return;
+        }
+    }
+
+    private function indexScopedPropertyAccess(Index $index, SplFileInfo $info, ScopedPropertyAccessExpression $node): void
+    {
+        $containerFqn = $node->scopeResolutionQualifier;
+
+        if (!$containerFqn instanceof QualifiedName) {
+            return;
+        }
+
+        $containerFqn = (string)$containerFqn->getResolvedName();
+        $memberName = $this->resolveScopedPropertyAccessName($node);
+
+        if (empty($memberName)) {
+            return;
+        }
+
+        $memberType = $this->resolveScopedPropertyAccessMemberType($node);
+
+        $this->writeIndex($index, $memberType, $containerFqn, $memberName, $info, $node);
+    }
+
+    public function beforeParse(Index $index, SplFileInfo $info): void
+    {
+        $fileRecord = $index->get(FileRecord::fromFileInfo($info));
+        assert($fileRecord instanceof FileRecord);
+
+        foreach ($fileRecord->references() as $outgoingReference) {
+            if ($outgoingReference->type() !== MemberRecord::RECORD_TYPE) {
+                continue;
+            }
+
+            $record = $index->get(MemberRecord::fromIdentifier($outgoingReference->identifier()));
+            assert($record instanceof MemberRecord);
+            $record->removeReference($fileRecord->identifier());
+            $index->write($record);
+        }
```

Comment:
Could abstract this, it's effectively copy-paste from the class references indexer